graphbolt/src/cuda/index_select_impl.cu

```diff
@@ -48,6 +53,283 @@ std::pair<torch::Tensor, torch::Tensor> Sort(
+    auto sub_indptr_pair = thrust::make_zip_iterator(
+        sub_indptr.data_ptr<indptr_t>(), sub_indptr_aligned.get());
+    thrust::tuple<indptr_t, indptr_t> zero_value{};
+    // Compute the prefix sum over actual and modified indegrees.
+    size_t workspace_size = 0;
+    CUDA_CALL(cub::DeviceScan::ExclusiveScan(
+        nullptr, workspace_size, modified_in_degree, sub_indptr_pair, PairSum{},
+        zero_value, num_nodes + 1, stream));
+    auto temp = allocator.AllocateStorage<char>(workspace_size);
+    CUDA_CALL(cub::DeviceScan::ExclusiveScan(
+        temp.get(), workspace_size, modified_in_degree, sub_indptr_pair,
+        PairSum{}, zero_value, num_nodes + 1, stream));
+  }
+  // Copy the modified number of edges.
+  CUDA_CALL(cudaMemcpyAsync(
+      &hop_size_aligned, sub_indptr_aligned.get() + num_nodes,
+      sizeof(hop_size_aligned), cudaMemcpyDeviceToHost, stream));
+  // Copy the actual total number of edges.
+  CUDA_CALL(cudaMemcpyAsync(
+      &hop_size, sub_indptr.data_ptr<indptr_t>() + num_nodes, sizeof(hop_size),
+      cudaMemcpyDeviceToHost, stream));
+  // synchronizes here, we can read hop_size and hop_size_aligned
+  CUDA_CALL(cudaStreamSynchronize(stream));
+  // Allocate output array with actual number of edges.
+  torch::Tensor sub_indices =
+      torch::empty(hop_size, nodes_options.dtype(indices.scalar_type()));
+  const dim3 block(BLOCK_SIZE);
+  const dim3 grid((hop_size_aligned + BLOCK_SIZE - 1) / BLOCK_SIZE);
+  // Perform the actual copying, of the indices array into
+  // sub_indices in an aligned manner.
+  CUDA_KERNEL_CALL(
+      _CSRRowWiseOneHopExtractorAlignedKernel, grid, block, 0, stream,
+      hop_size_aligned, num_nodes, sliced_indptr,
+      sub_indptr.data_ptr<indptr_t>(), sub_indptr_aligned.get(),
+      reinterpret_cast<indices_t*>(indices.data_ptr()),
+      reinterpret_cast<indices_t*>(sub_indices.data_ptr()), perm);
+  return {sub_indptr, sub_indices};
+}
+
+std::tuple<torch::Tensor, torch::Tensor> UVAIndexSelectCSCImpl(
+    torch::Tensor indptr, torch::Tensor indices, torch::Tensor nodes) {
+  // Sorting nodes so that accesses over PCI-e are more regular.
+  const auto perm_tensor =
+      Sort(nodes, cuda::NumberOfBits(indptr.size(0) - 1)).second;
+  auto stream = c10::cuda::getDefaultCUDAStream();
+  const int64_t num_nodes = nodes.size(0);
+
+  return AT_DISPATCH_INTEGRAL_TYPES(
+      indptr.scalar_type(), "UVAIndexSelectCSCIndptr", ([&] {
+        using indptr_t = scalar_t;
+        auto [in_degree_ptr, sliced_indptr_ptr] =
+            ComputeDegree(indptr.data_ptr<indptr_t>(), nodes, stream);
+        auto in_degree = in_degree_ptr.get();
+        auto sliced_indptr = sliced_indptr_ptr.get();
+        return GRAPHBOLT_DISPATCH_ELEMENT_SIZES(
+            indices.element_size(), "UVAIndexSelectCSCCopyIndices", ([&] {
+              return UVAIndexSelectCSCCopyIndices<indptr_t, element_size_t>(
+                  indices, sliced_indptr, num_nodes, in_degree,
```

Comment:
I would organize the parameters like: indices, num_nodes, in_degree, sliced_indptr. num_nodes relates to indices, and the order of in_degree and sliced_indptr matches the order of ComputeDegree's output.